src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1771,22 +1888,96 @@ public sealed class One : Tests
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        precondition: out Tests precondition,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    // Given T0 as a test that has already occurred, T1 as a subsequent test, and P as a pre-condition, we proceed as follow:
+                    //  - If T0=True implies T1, we rewrite T1 as (P ? True : T1) which reduces to (P || T1)
+                    //    Because we have determined that given T0=True, T1 would always succeed if P evaluates to True.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be further simplified as True which means T1 is insignificant.
+                    //  - If T0=True disallows T1, we rewrite T1 as (P ? False : T1) which reduces to (!P && T1)
+                    //    Because we have determined that given T0=True, T1 would never succeed if P evaluates to True.
+                    //    Note: If there is no pre-condition, i.e. P is True, the above will be further simplified as False which means T1 is impossible.
+                    //  - Otherwise, T1 is not rewritten -- the two are considered independent or there is no implication from one to the other.
```

Comment:
Thanks a lot. I was writing down some notes earlier this morning before seeing this update. It helped clarify my understanding. 
I'll share just in case.

CheckConsistentDecision returns with the following information:
A) P being true and T0 being true implies either: (1) T1 is true, (2) T1 is false, (3) nothing about T1.
B) P being true and T0 being false implies either: (1') T1 is true, (2') T1 is false, (3') nothing about T1.
C) If P is false, then T0 tells us nothing about T1.

In the branch where T0 was true, we're going to rewrite/simplify T1:
In situation (1) we rewrite T1 as "P or T1" (note this skips T1 if P was true).
In situation (2) we rewrite T1 as "!P and T1" (note this skips T1 is P was true).
In situation (3) neither T0 nor P tell us anything about T1, so we keep "T1".

In trivial cases where a precondition test wasn't useful, CheckConsistentDecision returns P as a True test and things simplify further:
In situation (1) we rewrite T1 as "True".
In situation (2) we rewrite T1 as "False".

In the branch where T0 was false, we do the same depending if we're in situation (1'), (2') or (3').